### ISSUE TITLE
[Close at end of milestone] general method of caching data

### ISSUE THREAD

**Issue by djayatillake** [lightdash/lightdash#353]
Proposal:

Lightdash should always store the results of a query as a temp table in the user DWH.

If the user then defines some user-defined dimensions or measures then the query needs to be rerun and the temp table recreated with these dimensions/measures as additional columns.

The UI should then query this temp table to show the results either selecting: 
columns in the temp table
AND if needed
some table calc fields like running totals etc which can be window functions or simpler run on the temp table

Advantages:
-If the user then wants to download the results and they are larger than what the UI can handle showing, then this is possible without re-running the query
-We can choose cloud storage as a destination to unload the results to relatively easily in the future
-We don't need to store any data on Lightdash infra and don't need to build architecture for caching in Lightdash, what is shown in the UI can just be a standard sql query on the temp table with a limit

Disadvantages:
-need to build a mapping between Lightdash query and temp table?
-need to decide/provide how to materialise temp tables and also how to dispose of them and when to
-need write permissions somewhere in the user DWH
-this probably can't be achieved while using dbt adaptor as method to connect to user DWHs

**Comment by djayatillake** [comment]
additional advantage from discussion with @owlas:

-temp table can be re-created automatically when dbt manifest metadata shows the queried tables have been updated and not before

option
-give user choice to clear cache (re-run temp table in the background) to get updated results (if for example a table had been updated outside of a dbt run)

**Comment by djayatillake** [comment]
Do we store a record per query in the pg db for a Lightdash instance?

Was thinking this could be an optional setting for an admin where they are willing to give Lightdash write access to a schema.

If the pg db currently stores the sql run for each query we could add an additional field which had the cache table location. Then on query run the table is checked for a query in the last x hours with the same sql, if it exists then take the results from the cache table, if not run the query and create the cache table, if the cache table exists and dbt freshness is old run the query and create the cache table, same again if user force reruns the cache. Go back to start of process each time cache is created or recreated.

Will need to add default limit 500 type button to explore to stop storing too much in the cache when users are just ecploring. Admin could set schema drop interval and it should default to every 24 hr at midnight if nothing or default is set.

**Comment by TuringLovesDeathMetal** [comment]
I don’t know where to put this, so I thought I’d drop it here:

Was reading through Looker’s table calculation docs and found them explaining a bit about how they run table calculations vs. metrics/dimensions:

https://docs.looker.com/exploring-data/using-table-calculations#:~:text=Looker%20expressions%20(Lexp).-,Table%20calculations%20are%20different%20from%20LookML%2Ddefined%20fields%20and%20custom%20fields,-There%20are%20a

I thought it was interesting because it looks like they cache the data they return from the query, and you can only apply table calculations on the returned dataset.

I think this is what we were talking about above, but I could never find it explained in Looker's docs anywhere

**Comment by 12ian34** [comment]
Has been requested twice recently by prospects

**Comment by TuringLovesDeathMetal** [comment]
@12ian34 were the requests because of pricing? Or because of performance (speed)?

**Comment by 12ian34** [comment]
For one prospect (large company) it was about speed and pricing. For the other (small start-up) it was about speed.

**Comment by olly-writes-code** [comment]
At my company we pretty do the whole caching ourselves due to slow query speed and lots of data sources. Not many companies can quickly pivot away from redshift, snowflake, big query just because the query speed is bad for a web viz. Hence, the effort on caching. I know this isn't easy though, but it is something Looker do. While doing some research I came across [this](https://github.com/cube-js/cube.js) and starting of dreaming of combining it with lightdash 😈

**Comment by djayatillake** [comment]
One hopes that the dbt server will have a cache built in, but not clear yet. Lightdash folks may know this as they are closer to it.

**Comment by ZeRego** [comment]
Leaving here an idea from a slack thread
https://getdbt.slack.com/archives/C026WJE4A69/p1672889282568499?thread_ts=1672295558.669019&cid=C026WJE4A69

> Mike Weinberg
In snowflake, the moment you run a query, you can reference its result set - lightdash could retrieve query id’s after running in order to access the cached result, and then at least in snowflake you could call TABLE on it and run queries against it for defining user defined dimensions, such that the marginal cost of compute is just the scan of the result set plus the small additional cost of the extra fields. That would be quite fast.

> Mike Weinberg
 among other useful doc pages, this is a good place to start: https://docs.snowflake.com/en/user-guide/querying-persisted-results.html
Feel free to dm me if you wanna schedule time to talk about this - improving lightdash is worth my time.

**Comment by owlas** [comment]
Some good triggers for caching policies will be familiar to looker users:
- refresh cache on a regular schedule
- refresh cache based on when the data changes

https://cloud.google.com/looker/docs/reference/param-model-datagroup

**Comment by rephus** [comment]
I agree caching might be valuable, specially if a chart/dashboard doesn't change very often. However, implementing caching has many implications, not only is hard to implement technically (how often do we cache, when do we invalidate caches?), but this  also means we will have to store the customer's data, which could raise some security concerns, specially on cloud. We will have to implement this eventually though, but it is not in scope right now.

Some people might want to investigate their own internal warehouse caching first. I know some warehouses support this internally, like bigquery:  https://cloud.google.com/bigquery/docs/cached-results

**Comment by 12ian34** [comment]
A prospect has asked for caching for embedding - created a separate issue here: https://github.com/lightdash/lightdash/issues/12275

**Comment by ZeRego** [comment]
Closing this ticket as we have a caching mechanism. Documentation available here https://docs.lightdash.com/references/caching
